File: src/ai/tools.py
"""AI工具函数 - 让AI可以操作程序数据"""

import json
import time


class CardTools:
    """卡片操作工具集"""

    def __init__(self, app):
        """初始化工具集

        app: PapyrusApp实例
        """
        self.app = app

    def _log_event(self, event_type, data=None, level="INFO"):
        """安全地写入事件日志（logger 可能为 None）"""
        logger = getattr(self.app, "logger", None)
        if logger and hasattr(logger, "log_event"):
            logger.log_event(event_type, data, level=level)

    def _try_refresh_ui(self):
        """尝试刷新主学习区"""
        try:
            if hasattr(self.app, "next_card"):
                self.app.next_card()
        except Exception:
            pass

    def get_tools_definition(self):
        """返回工具定义，供AI理解"""
        return """
你可以使用以下工具来操作学习卡片：

1. create_card(question, answer, tags=[])
   - 创建新卡片
   - 参数：question(题目), answer(答案), tags(标签列表，可选)

2. update_card(card_index, question=None, answer=None)
   - 更新指定卡片
   - 参数：card_index(卡片索引), question(新题目), answer(新答案)

3. delete_card(card_index)
   - 删除指定卡片
   - 参数：card_index(卡片索引)

4. search_cards(keyword)
   - 搜索包含关键词的卡片
   - 参数：keyword(搜索关键词)
   - 返回：匹配的卡片列表

5. get_card_stats()
   - 获取学习统计
   - 返回：总卡片数、待复习数、平均难度等

6. generate_practice_set(topic, count=5)
   - 根据主题生成练习题（当前版本未实现；会返回未实现错误）
   - 参数：topic(主题), count(数量)


使用格式：
```json
{
  "tool": "create_card",
  "params": {
    "question": "什么是递归？",
    "answer": "递归是函数调用自身的过程"
  }
}
```

注意：所有修改操作会立即执行并保存。
"""

    def create_card(self, question, answer, tags=None):
        """创建新卡片并立即保存"""
        if not question or not answer:
            return {"success": False, "error": "题目和答案不能为空"}

        new_card = {
            "q": question,
            "a": answer,
            "next_review": 0,
            "interval": 0,
            "tags": tags or [],
        }

        self.app.cards.append(new_card)
        self.app.save_data()
        self._log_event("tool.create_card", {"question": question[:50]})
        self._try_refresh_ui()

        return {
            "success": True,
            "message": "卡片已创建并保存",
            "card": new_card,
        }

    def update_card(self, card_index, question=None, answer=None):
        """更新卡片并立即保存"""
        if card_index < 0 or card_index >= len(self.app.cards):
            return {"success": False, "error": "卡片索引无效"}

        card = self.app.cards[card_index]
        old_q, old_a = card["q"], card["a"]

        if question:
            card["q"] = question
        if answer:
            card["a"] = answer

        if not question and not answer:
            return {"success": False, "error": "没有提供更新内容"}

        self.app.save_data()
        self._log_event("tool.update_card", {"index": card_index})
        self._try_refresh_ui()

        return {
            "success": True,
            "message": "卡片已更新并保存",
            "old": {"q": old_q, "a": old_a},
            "new": {"q": card["q"], "a": card["a"]},
        }

    def delete_card(self, card_index):
        """删除卡片并立即保存"""
        if card_index < 0 or card_index >= len(self.app.cards):
            return {"success": False, "error": "卡片索引无效"}

        card = self.app.cards.pop(card_index)
        self.app.save_data()
        self._log_event("tool.delete_card", {"index": card_index, "question": card["q"][:50]})
        self._try_refresh_ui()

        return {
            "success": True,
            "message": "卡片已删除并保存",
            "deleted_card": card,
        }

    def search_cards(self, keyword):
        """搜索卡片"""
        keyword = (keyword or "").lower()
        results = []

        for i, card in enumerate(self.app.cards):
            q = card.get("q", "")
            ans = card.get("a", "")
            if keyword in q.lower() or keyword in ans.lower():
                results.append(
                    {
                        "index": i,
                        "question": q,
                        "answer": ans[:100] + "..." if len(ans) > 100 else ans,
                    }
                )

        return {
            "success": True,
            "count": len(results),
            "results": results,
        }

    def get_card_stats(self):
        """获取统计信息"""
        total = len(self.app.cards)
        due = len(self.app.get_due_cards())

        efs = [c.get("ef", 2.5) for c in self.app.cards]
        avg_ef = sum(efs) / len(efs) if efs else 2.5

        reps = [c.get("repetitions", 0) for c in self.app.cards]

        return {
            "success": True,
            "stats": {
                "total_cards": total,
                "due_cards": due,
                "average_ef": round(avg_ef, 2),
                "max_repetitions": max(reps) if reps else 0,
                "cards_mastered": len([r for r in reps if r >= 5]),
            },
        }

    def execute_tool(self, tool_name, params):
        """执行工具调用"""
        tools = {
            "create_card": self.create_card,
            "update_card": self.update_card,
            "delete_card": self.delete_card,
            "search_cards": self.search_cards,
            "get_card_stats": self.get_card_stats,
            "generate_practice_set": self.generate_practice_set,
        }

        if tool_name not in tools:
            self._log_event("tool.unknown", {"tool": tool_name}, level="WARNING")
            return {"success": False, "error": f"未知工具: {tool_name}"}

        self._log_event("tool.execute_start", {"tool": tool_name, "params": params})
        start = time.time()
        try:
            result = tools[tool_name](**params)
            elapsed = round(time.time() - start, 4)

            if not isinstance(result, dict):
                result = {"success": True, "result": result}

            self._log_event(
                "tool.execute_ok",
                {
                    "tool": tool_name,
                    "elapsed_s": elapsed,
                    "success": result.get("success"),
                },
            )
            return result

        except Exception as e:
            elapsed = round(time.time() - start, 4)
            self._log_event(
                "tool.execute_error",
                {
                    "tool": tool_name,
                    "elapsed_s": elapsed,
                    "error": str(e),
                },
                level="ERROR",
            )
            return {"success": False, "error": str(e)}

    def generate_practice_set(self, topic, count=5):
        """根据主题生成练习题（占位：未实现）"""
        return {
            "success": False,
            "error": "generate_practice_set 当前版本未实现",
            "topic": topic,
            "count": count,
        }

    def parse_tool_call(self, ai_response):
        """从AI响应中解析工具调用"""
        import re

        json_pattern = r"```json\s*(\{.*?\})\s*```"
        matches = re.findall(json_pattern, ai_response, re.DOTALL)

        if not matches:
            return None

        try:
            tool_call = json.loads(matches[0])
            if "tool" in tool_call and "params" in tool_call:
                self._log_event(
                    "tool.parse_ok",
                    {
                        "tool": tool_call["tool"],
                        "params": tool_call["params"],
                    },
                )
                return tool_call
        except Exception as e:
            self._log_event("tool.parse_error", {"error": str(e)}, level="WARNING")

        return None

File: backend/src/ai/tools.ts
import { randomUUID } from 'node:crypto';
import {
  loadAllCards,
  insertCard,
  updateCard as dbUpdateCard,
  deleteCardById,
  getCardsDueBefore,
  getCardCount,
} from '../db/database.js';
import type { CardRecord } from '../core/types.js';
import type { PapyrusLogger } from '../utils/logger.js';

export interface ToolResult {
  success: boolean;
  error?: string;
  message?: string;
  card?: CardRecord;
  old?: { q: string; a: string };
  new?: { q: string; a: string };
  deleted_card?: CardRecord;
  count?: number;
  results?: Array<{ index: number; question: string; answer: string }>;
  stats?: {
    total_cards: number;
    due_cards: number;
    average_ef: number;
    max_repetitions: number;
    cards_mastered: number;
  };
  topic?: string;
}

export interface ToolCall {
  tool: string;
  params: Record<string, unknown>;
}

export interface OpenAIToolDef {
  type: 'function';
  function: {
    name: string;
    description: string;
    parameters: {
      type: 'object';
      properties: Record<string, {
        type: 'string' | 'integer' | 'number' | 'boolean' | 'array' | 'object';
        description?: string;
        items?: { type: 'string' | 'integer' | 'number' | 'boolean' };
      }>;
      required?: string[];
    };
  };
}

export interface ParsedAIResponse {
  content: string;
  reasoning: string | null;
  tool_call: ToolCall | null;
}

function safeFloat(value: unknown, defaultValue: number): number {
  if (typeof value === 'number') return value;
  if (typeof value === 'string') {
    const parsed = parseFloat(value);
    if (!Number.isNaN(parsed)) return parsed;
  }
  return defaultValue;
}

function safeInt(value: unknown, defaultValue: number): number {
  if (typeof value === 'number' && Number.isInteger(value)) return value;
  if (typeof value === 'string') {
    const parsed = parseInt(value, 10);
    if (!Number.isNaN(parsed)) return parsed;
  }
  return defaultValue;
}

export class CardTools {
  private logger: PapyrusLogger | null;

  constructor(logger?: PapyrusLogger) {
    this.logger = logger ?? null;
  }

  private logEvent(eventType: string, data: unknown = null, level = 'INFO'): void {
    this.logger?.logEvent(eventType, data, level);
  }

  getToolsDefinition(): string {
    return `
你可以使用以下工具来操作学习卡片：

1. 创建卡片 (create_card)
参数：
- question: 题目内容
- answer: 答案内容
- tags: 标签列表（可选）
示例：{"tool": "create_card", "params": {"question": "...", "answer": "...", "tags": ["数学"]}}

2. 更新卡片 (update_card)
参数：
- card_index: 卡片索引
- question: 新题目（可选）
- answer: 新答案（可选）
示例：{"tool": "update_card", "params": {"card_index": 0, "question": "..."}}

3. 删除卡片 (delete_card)
参数：
- card_index: 卡片索引
示例：{"tool": "delete_card", "params": {"card_index": 0}}

4. 搜索卡片 (search_cards)
参数：
- keyword: 搜索关键词
示例：{"tool": "search_cards", "params": {"keyword": "数学"}}

5. 获取统计 (get_card_stats)
无参数
示例：{"tool": "get_card_stats", "params": {}}

6. 生成练习集 (generate_practice_set)
参数：
- topic: 主题
- count: 题目数量（可选，默认5）
示例：{"tool": "generate_practice_set", "params": {"topic": "三角函数", "count": 5}}

使用格式：
\`\`\`json
{"tool": "工具名", "params": {...}}
\`\`\`

注意：所有修改操作会立即执行并保存。
`;
  }

  getToolsForOpenAI(): OpenAIToolDef[] {
    return [
      {
        type: 'function',
        function: {
          name: 'create_card',
          description: '创建一张新的学习卡片并立即保存。用于用户希望记录知识点、问答对或复习内容时',
          parameters: {
            type: 'object',
            properties: {
              question: { type: 'string', description: '题目内容' },
              answer: { type: 'string', description: '答案内容' },
              tags: { type: 'array', description: '标签列表', items: { type: 'string' } },
            },
            required: ['question', 'answer'],
          },
        },
      },
      {
        type: 'function',
        function: {
          name: 'update_card',
          description: '根据卡片索引更新已存在的卡片。仅传入需要修改的字段',
          parameters: {
            type: 'object',
            properties: {
              card_index: { type: 'integer', description: '卡片在列表中的索引（从 0 开始）' },
              question: { type: 'string', description: '新的题目（可选）' },
              answer: { type: 'string', description: '新的答案（可选）' },
            },
            required: ['card_index'],
          },
        },
      },
      {
        type: 'function',
        function: {
          name: 'delete_card',
          description: '根据索引删除一张卡片',
          parameters: {
            type: 'object',
            properties: {
              card_index: { type: 'integer', description: '卡片在列表中的索引（从 0 开始）' },
            },
            required: ['card_index'],
          },
        },
      },
      {
        type: 'function',
        function: {
          name: 'search_cards',
          description: '在题目、答案、标签中搜索关键词，返回匹配的卡片列表',
          parameters: {
            type: 'object',
            properties: {
              keyword: { type: 'string', description: '搜索关键词' },
            },
            required: ['keyword'],
          },
        },
      },
      {
        type: 'function',
        function: {
          name: 'get_card_stats',
          description: '获取卡片库的整体统计：总数、到期数、平均熟练度、最高复习次数、已掌握卡片数',
          parameters: {
            type: 'object',
            properties: {},
            required: [],
          },
        },
      },
      {
        type: 'function',
        function: {
          name: 'generate_practice_set',
          description: '基于主题生成一组练习卡片',
          parameters: {
            type: 'object',
            properties: {
              topic: { type: 'string', description: '主题' },
              count: { type: 'integer', description: '题目数量，默认 5' },
            },
            required: ['topic'],
          },
        },
      },
    ];
  }

  createCard(question: string, answer: string, tags?: string[]): ToolResult {
    if (!question || !answer) {
      return { success: false, error: '题目和答案不能为空' };
    }

    const newCard: CardRecord = {
      id: randomUUID(),
      q: question,
      a: answer,
      next_review: Date.now() / 1000,
      interval: 0,
      tags: tags ?? [],
      ef: 2.5,
      repetitions: 0,
    };

    insertCard(newCard, this.logger ?? undefined);
    this.logEvent('tool.create_card', { question: question.slice(0, 50) });

    return {
      success: true,
      message: '卡片已创建并保存',
      card: newCard,
    };
  }

  updateCard(cardIndex: number, question?: string, answer?: string): ToolResult {
    const cards = loadAllCards(this.logger ?? undefined);
    if (cardIndex < 0 || cardIndex >= cards.length) {
      return { success: false, error: '卡片索引无效' };
    }

    const card = cards[cardIndex];
    if (!card) {
      return { success: false, error: '卡片索引无效' };
    }

    const oldQ = card.q;
    const oldA = card.a;

    if (question) card.q = question;
    if (answer) card.a = answer;

    dbUpdateCard(card, this.logger ?? undefined);
    this.logEvent('tool.update_card', {
      index: cardIndex,
      old_q: oldQ.slice(0, 50),
      new_q: question?.slice(0, 50),
    });

    return {
      success: true,
      message: '卡片已更新并保存',
      old: { q: oldQ, a: oldA },
      new: { q: card.q, a: card.a },
    };
  }

  deleteCard(cardIndex: number): ToolResult {
    const cards = loadAllCards(this.logger ?? undefined);
    if (cardIndex < 0 || cardIndex >= cards.length) {
      return { success: false, error: '卡片索引无效' };
    }

    const card = cards[cardIndex];
    if (!card) {
      return { success: false, error: '卡片索引无效' };
    }

    deleteCardById(card.id, this.logger ?? undefined);
    this.logEvent('tool.delete_card', { index: cardIndex, question: card.q.slice(0, 50) });

    return {
      success: true,
      message: '卡片已删除并保存',
      deleted_card: card,
    };
  }

  searchCards(keyword: string): ToolResult {
    const keywordLower = (keyword || '').toLowerCase();
    const cards = loadAllCards(this.logger ?? undefined);
    const results: Array<{ index: number; question: string; answer: string }> = [];

    for (let i = 0; i < cards.length; i++) {
      const card = cards[i];
      if (!card) continue;
      const q = card.q;
      const a = card.a;
      if (q.toLowerCase().includes(keywordLower) ||
          a.toLowerCase().includes(keywordLower) ||
          card.tags.some(t => t.toLowerCase().includes(keywordLower))) {
        results.push({
          index: i,
          question: q,
          answer: a.length > 100 ? `${a.slice(0, 100)}...` : a,
        });
      }
    }

    return {
      success: true,
      message: `找到 ${results.length} 张相关卡片`,
      count: results.length,
      results,
    };
  }

  getCardStats(): ToolResult {
    const cards = loadAllCards(this.logger ?? undefined);
    const total = cards.length;
    const now = Date.now() / 1000;
    const due = getCardsDueBefore(now).length;

    const efs = cards.map(c => safeFloat(c.ef, 2.5));
    const avgEf = efs.length > 0 ? efs.reduce((a, b) => a + b, 0) / efs.length : 2.5;
    const reps = cards.map(c => safeInt(c.repetitions, 0));

    return {
      success: true,
      stats: {
        total_cards: total,
        due_cards: due,
        average_ef: Math.round(avgEf * 100) / 100,
        max_repetitions: reps.length > 0 ? Math.max(...reps) : 0,
        cards_mastered: reps.filter(r => r >= 5).length,
      },
    };
  }

  executeTool(toolName: string, params: Record<string, unknown>): ToolResult {
    const knownTools = new Set([
      'create_card',
      'update_card',
      'delete_card',
      'search_cards',
      'get_card_stats',
      'generate_practice_set',
    ]);

    if (!knownTools.has(toolName)) {
      this.logEvent('tool.unknown', { tool: toolName }, 'WARNING');
      return { success: false, error: `未知工具: ${toolName}` };
    }

    this.logEvent('tool.execute_start', { tool: toolName, params });
    const start = Date.now();

    try {
      let result: ToolResult;

      if (toolName === 'create_card') {
        const question = params.question;
        const answer = params.answer;
        const tags = params.tags;
        if (typeof question !== 'string' || typeof answer !== 'string') {
          return { success: false, error: 'question 和 answer 必须是字符串' };
        }
        const tagList = Array.isArray(tags) ? tags.map(t => String(t)) : undefined;
        result = this.createCard(question, answer, tagList);
      } else if (toolName === 'update_card') {
        const cardIndex = params.card_index;
        const question = params.question;
        const answer = params.answer;
        if (typeof cardIndex !== 'number') {
          return { success: false, error: 'card_index 必须是整数' };
        }
        result = this.updateCard(
          Math.floor(cardIndex),
          typeof question === 'string' ? question : undefined,
          typeof answer === 'string' ? answer : undefined,
        );
      } else if (toolName === 'delete_card') {
        const cardIndex = params.card_index;
        if (typeof cardIndex !== 'number') {
          return { success: false, error: 'card_index 必须是整数' };
        }
        result = this.deleteCard(Math.floor(cardIndex));
      } else if (toolName === 'search_cards') {
        const keyword = params.keyword;
        if (typeof keyword !== 'string') {
          return { success: false, error: 'keyword 必须是字符串' };
        }
        result = this.searchCards(keyword);
      } else if (toolName === 'get_card_stats') {
        result = this.getCardStats();
      } else {
        const topic = params.topic;
        const count = params.count;
        if (typeof topic !== 'string') {
          return { success: false, error: 'topic 必须是字符串' };
        }
        result = this.generatePracticeSet(topic, typeof count === 'number' ? count : 5);
      }

      const elapsed = (Date.now() - start) / 1000;
      this.logEvent('tool.execute_ok', {
        tool: toolName,
        elapsed_s: elapsed,
        result_type: result.success ? 'success' : 'error',
      });
      return result;
    } catch (exc) {
      const elapsed = (Date.now() - start) / 1000;
      this.logEvent(
        'tool.execute_error',
        { tool: toolName, elapsed_s: elapsed, error: exc instanceof Error ? exc.message : String(exc) },
        'ERROR',
      );
      return { success: false, error: exc instanceof Error ? exc.message : String(exc) };
    }
  }

  generatePracticeSet(topic: string, count = 5): ToolResult {
    return {
      success: false,
      error: '生成功能尚未实现',
      topic,
      count,
    };
  }

  parseToolCall(aiResponse: string): ToolCall | null {
    const pattern = /```json\s*(\{.*?\})\s*```/s;
    const matches = aiResponse.match(pattern);
    if (!matches || !matches[1]) return null;

    try {
      const toolCallObj = JSON.parse(matches[1]) as unknown;
      if (
        toolCallObj !== null &&
        typeof toolCallObj === 'object' &&
        typeof (toolCallObj as Record<string, unknown>).tool === 'string' &&
        typeof (toolCallObj as Record<string, unknown>).params === 'object'
      ) {
        const tc = toolCallObj as Record<string, unknown>;
        const toolCall: ToolCall = {
          tool: tc.tool as string,
          params: tc.params as Record<string, unknown>,
        };
        this.logEvent('tool.parse_ok', {
          tool: toolCall.tool,
          params_keys: Object.keys(toolCall.params),
        });
        return toolCall;
      }
    } catch (exc) {
      this.logEvent('tool.parse_error', { error: exc instanceof Error ? exc.message : String(exc) }, 'WARNING');
    }
    return null;
  }
}

export class AIResponseParser {
  private static readonly REASONING_TAGS: Array<[RegExp, string]> = [
    [/<think>(.*?)<\/think>/gs, 'think'],
    [/<reasoning>(.*?)<\/reasoning>/gs, 'reasoning'],
    [/<thought>(.*?)<\/thought>/gs, 'thought'],
  ];

  private static readonly TOOL_CALL_PATTERNS = [
    /```json\s*(\{.*?\})\s*```/gs,
    /```\s*(\{[^{}]*"tool"[^{}]*\})\s*```/gs,
  ];

  static parseReasoning(content: string): { cleaned: string; reasoning: string | null } {
    const reasoningParts: string[] = [];
    let cleanedContent = content;

    for (const [pattern] of this.REASONING_TAGS) {
      const matches = [...content.matchAll(pattern)];
      if (matches.length > 0) {
        for (const match of matches) {
          const text = match[1];
          if (text) reasoningParts.push(text.trim());
        }
        cleanedContent = cleanedContent.replace(pattern, '');
      }
    }

    cleanedContent = cleanedContent.replace(/\n{3,}/g, '\n\n').trim();
    const reasoning = reasoningParts.length > 0 ? reasoningParts.join('\n\n') : null;
    return { cleaned: cleanedContent, reasoning };
  }

  static parseToolCall(content: string): ToolCall | null {
    for (const pattern of this.TOOL_CALL_PATTERNS) {
      const matches = [...content.matchAll(pattern)];
      for (const match of matches) {
        try {
          const text = match[1];
          if (!text) continue;
          const obj = JSON.parse(text) as unknown;
          if (
            obj !== null &&
            typeof obj === 'object' &&
            typeof (obj as Record<string, unknown>).tool === 'string' &&
            typeof (obj as Record<string, unknown>).params === 'object'
          ) {
            return {
              tool: (obj as Record<string, unknown>).tool as string,
              params: (obj as Record<string, unknown>).params as Record<string, unknown>,
            };
          }
        } catch {
          continue;
        }
      }
    }
    return null;
  }

  static parseResponse(response: string, reasoningContent?: string | null): ParsedAIResponse {
    const { cleaned, reasoning } = this.parseReasoning(response);
    const finalReasoning = reasoningContent && !reasoning ? reasoningContent : reasoning;
    const toolCall = this.parseToolCall(cleaned) ?? this.parseToolCall(response);
    const contentWithoutTools = this.removeToolCallMarkers(cleaned);

    return {
      content: contentWithoutTools,
      reasoning: finalReasoning,
      tool_call: toolCall,
    };
  }

  static removeToolCallMarkers(content: string): string {
    let cleaned = content;
    for (const pattern of this.TOOL_CALL_PATTERNS) {
      cleaned = cleaned.replace(pattern, '');
    }
    cleaned = cleaned.replace(/\n{2,}/g, '\n').trim();
    return cleaned;
  }
}

export const parseToolCall = AIResponseParser.parseToolCall;
export const parseReasoning = AIResponseParser.parseReasoning;
export const parseResponse = AIResponseParser.parseResponse;
